### PR TITLE
[Model] Add Preconditioned Gated DeltaNet (PGDN) and KDA (PKDA)

### PR DESCRIPTION
**Pull request by ntumm120**
## Summary

Adds curvature-aware **preconditioned** variants of Gated DeltaNet and KDA from
*Preconditioned DeltaNet: Curvature-aware Sequence Modeling for Linear Recurrences*
([arXiv:2604.21100](https://arxiv.org/abs/2604.21100), ICML 2026).

Reference implementation / training code (paper repo): https://github.com/ntumm120/preconditioned-deltanet

**Ops**
- `fla/ops/atk/` — shared ATK preconditioner kernels (fwd/bwd)
- `fla/ops/precond_gated_delta_rule/` — PGDN chunk + fused_recurrent
- `fla/ops/precond_kda/` — PKDA chunk (+ intra) + fused_recurrent

**Layers** — `fla/layers/precond_gated_deltanet.py`, `fla/layers/precond_kda.py`
**Models** — `fla/models/precond_gated_deltanet/`, `fla/models/precond_kda/`
**Tests** — op + model tests for both

Registered in `fla/{__init__,layers/__init__,models/__init__,ops/__init__}.py` and the
README model table, following existing conventions (cf. #939).

All kernels are written with manual pointer arithmetic + masks (no `tl.make_block_ptr`),
matching the current upstream KDA/GDN kernels so they build against Triton main.

### Note on `FLACache`

`FLACache`/`FLALayer` gain an **optional, backward-compatible** `A_state` field
(`fla/models/utils.py`) to carry the preconditioner's auxiliary recurrent state during
generation. It defaults to `None` and is a no-op for all existing models.

### Backends

Kernels are **Triton-only**. The shared `chunk_bwd_dqkwg` path reuses the existing backend
dispatch, so PGDN inherits the same Hopper/tilelang requirement as baseline GDN/KDA (#640).
FlashKDA/TileLang precond backends are possible follow-ups.

## Test plan

Full precond op + model suite on A100 (`FLA_DISABLE_BACKEND_DISPATCH=1`):

```
pytest tests/ops/test_precond_kda.py tests/ops/test_precond_gated_delta.py
pytest tests/models/test_modeling_precond_kda.py tests/models/test_modeling_precond_gated_deltanet.py
```

→ **78 op tests + 7 model tests pass, 0 failed** (dense + varlen + prefill, `safe_gate` on/off,
transpose-state, `disable_recompute`, fp16/bf16/fp32). GVA (`num_v_heads > num_heads`) path stays
finite in loss and gradients. `ruff` clean.

## Benchmark (kernel changes only)

A100-80GB, bf16, `B=4, T=4096, H=4, D=128`. "baseline" is the non-preconditioned op
(`chunk_kda` / `chunk_gated_delta_rule`); "precond" is the new op — the ratio is the cost of the
added curvature-aware preconditioning step, which is the intended overhead:

| op          | shape  | pass    | baseline | precond | ratio |
|-------------|--------|---------|---------:|--------:|------:|
| precond_kda | dense  | fwd     | 0.874 ms | 1.397 ms | 1.60x |
| precond_kda | dense  | fwd+bwd | 4.796 ms | 6.784 ms | 1.41x |
| precond_kda | varlen | fwd     | 0.868 ms | 1.746 ms | 2.01x |
| precond_kda | varlen | fwd+bwd | 5.029 ms | 7.456 ms | 1.48x |
| precond_gdn | dense  | fwd     | 0.634 ms | 1.188 ms | 1.87x |
| precond_gdn | dense  | fwd+bwd | 2.944 ms | 5.068 ms | 1.72x |
| precond_gdn | varlen | fwd     | 0.681 ms | 1.582 ms | 2.32x |
| precond_gdn | varlen | fwd+bwd | 3.369 ms | 6.377 ms | 1.89x |

Median of 50 runs after 10 warmups.

## Breaking changes

None. The only shared-code change is the optional `A_state` field on `FLACache`/`FLALayer`,
which defaults to `None` and does not affect any existing model.

## Checklist

- [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md) and follow its conventions (code style, docstrings, commit prefixes).
- [x] I have read [AGENTS.md](../AGENTS.md) and, where my change matches its scope, the relevant skill under [.agents/skills](../.agents/skills).
- [x] This is not a minor/cosmetic-only PR (typo, formatting, style-only tweaks).
- [x] Dependent tests pass locally or in CI; new behavior is covered by tests where applicable.
- [x] Kernel changes include same-hardware before/after benchmark numbers (dense + varlen where applicable).

### CODE REVIEW

File: fla/layers/precond_kda.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        g = self.f_proj(hidden_states)
+        beta = self.b_proj(hidden_states).sigmoid()
+
+        q, k, g = (rearrange(x, "... (h d) -> ... h d", d=self.head_k_dim) for x in (q, k, g))
+        v = rearrange(v, "... (h d) -> ... h d", d=self.head_v_dim)
+
+        # for multi-value attention, we repeat the inputs for simplicity.
+        if self.num_v_heads > self.num_heads:
+            q, k, g = (repeat(x, "... h d -> ... (h g) d", g=self.num_v_heads // self.num_heads) for x in (q, k, g))
+            beta = repeat(beta, "... h -> ... (h g)", g=self.num_v_heads // self.num_heads)
```
[the remark this answers is not in the document]
Comment (ntumm120) [reply fla/layers/precond_kda.py:292]: Good catch, fixed in the latest push. `g` already comes out of `f_proj` at `num_v_heads`, so it shouldn't be in the repeat — now only `q`/`k`/`beta` get expanded to `num_v_heads`.

File: fla/models/precond_kda/modeling_precond_kda.py [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+                    else:
+                        module.A_log.copy_(nn.init.uniform_(module.A_log, a=a_log_lower, b=16).log())
+                module.A_log._no_weight_decay = True
+                if not getattr(module.dt_bias, "_is_hf_initialized", False):
+                    dt = torch.exp(
+                        nn.init.uniform_(module.dt_bias) * (math.log(0.1) - math.log(0.001)) + math.log(0.001),
+                    ).clamp(min=1e-4)
+                    inv_dt = dt + torch.log(-torch.expm1(-dt))
+                    module.dt_bias.copy_(inv_dt)
+                module.dt_bias._is_hf_initialized = True
```
[the remark this answers is not in the document]
Comment (ntumm120) [reply fla/models/precond_kda/modeling_precond_kda.py:165]: This one's intentional — it mirrors baseline KDA. `dt_bias` here feeds `fused_kda_gate` exactly like in KDA, and KDA's `_init_weights` also copies `inv_dt` into it rather than zeroing it. Zeroing would diverge from KDA's gate init, so I'm keeping it consistent.

File: fla/models/precond_kda/modeling_precond_kda.py [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+            past_key_values=past_key_values,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            **kwargs,
+        )
+
+        hidden_states = outputs[0]
+        fuse_linear_and_cross_entropy = self.config.fuse_cross_entropy and self.training and labels is not None
```
[the remark this answers is not in the document]
Comment (ntumm120) [reply fla/models/precond_kda/modeling_precond_kda.py:370]: This matches baseline KDA and the rest of fla — `fuse_linear_and_cross_entropy` is gated on `fuse_cross_entropy` across the models, and the config doesn't carry a separate `fuse_linear_cross_entropy` flag. I'd rather stay consistent with the other models here; if you want to split that flag I think it's better done repo-wide in its own PR.

**Comment by ntumm120** [comment]
@yzhangcs @sustcsonglin @zhiyuan1i lmk if there's anything I need to change

**Comment by ntumm120** [comment]
@zhiyuan1i Can you please look at this?

**Comment by ntumm120** [comment]
@zhiyuan1i is this ever gonna get merged?

**Review by zhiyuan1i — CHANGES_REQUESTED** [review]
Thanks for the solid work — the core math checks out on the H == HV path (fwd/bwd verified against the naive reference), CI is green, and the integration follows repo conventions. But there are several correctness gaps in places the tests don't reach, so this needs another round before merging:

1. GVA is silently wrong. `fla/ops/precond_kda/chunk_precond_kkt_fwd.py:70-71` reads `g`/`beta` with H-strides while they are `[B, T, HV]` tensors (and A is allocated one head-dimension short at :141); the PGDN layer never repeats q/k, so `num_v_heads > num_heads` is directly reachable and computes garbage, while the docstring advertises GVA support. Please either implement GVA properly or reject it loudly.

2. The final A-state gradient is silently dropped. Both backward functions take `dat` but return no gradient for the initial A state (`fla/ops/precond_kda/chunk.py:456`, `fla/ops/precond_gated_delta_rule/chunk.py:359`).

3. `cp_context` is advertised but not implemented — the fwd/bwd lacks the baseline's pre_process / compress_h0 / expand_h0 plumbing, so CP silently computes wrong results. Please implement it or raise for now.

4. `fla/ops/precond_kda/fused_recurrent.py:191` loads `g` without a mask — the same out-of-bounds read that #1135 just fixed in the KDA kernel, re-introduced here.

5. `LegacyFLACache.update` isn't adapted for the A state: on transformers ≤ 4.53.3 (allowed by pyproject), generation with use_cache raises TypeError on the first step.

Worth fixing in the same round: the A state is untested end-to-end (no test asserts `at`, passes a nonzero `initial_A_state`, or verifies `dh0_atk`); int64 index guards regressed vs the baseline in 8+ kernels (ATK included); ATK's docstring and kernel disagree on the H/HV contract; and the new ops are not registered in `benchmarks/ops/registry.py` per CONTRIBUTING. Since this PR is introducing new public knobs, please also pin down the init semantics now (the `a_log_init_lower` hidden knob shifts the KDA init distribution off the baseline; PGDN's `use_exp2` behavior) — changing them after merge would be breaking.

**Comment by ntumm120** [comment]
@zhiyuan1i Thanks for the review — all five points you made were addressed in the latest push:

1. **GVA**: implemented properly. The KKT kernel now handles `HV > H` (`i_hv // (HV // H)` key-head indexing), and the ATK preconditioner is pinned to the key-head (`H`) contract throughout. New `test_chunk_gva` cases verify forward + all gradients against the naive reference at 2x/4x group sizes.
2. **A-state gradient**: `dat` now propagates — injected into the ATK reverse-scan carry at the final chunk (`at == ac[:, last_chunk]`). New `test_chunk_A_state` cases for both ops keep `at` in the loss and check `at`/`dh0_atk` against autograd through the naive reference.
3. **CP**: raises `NotImplementedError` instead of computing wrong results; proper support is a follow-up.
4. **OOB load**: fixed, plus two more of the same class (`gk`/`gv` in the PGDN fused_recurrent).
5. **LegacyFLACache**: `update` now handles `A_state`.

The other items too: int64 guards restored across the precond/ATK kernels (baseline `program_id(...).to(tl.int64)` idiom), both ops registered in `benchmarks/ops/registry.py`, the A-state now tested end-to-end per point 2, and the ATK H/HV docstring mismatch resolved by pinning ATK to the key-head contract per point 1. On init semantics and `use_exp2`: agreed on both, landed now — `a_log_init_lower` is a declared config knob with each model defaulting to its baseline's init (PrecondKDA -> `uniform(1, 16)` like KDA; PGDN keeps `uniform(0, 16)` like GDN), and `use_exp2` is removed from the PGDN path entirely, finishing the repo-wide exp2 migration that KDA/PKDA already follow. The branch is also synced with current main.

Re-validated on A100: 85 op + 7 model tests, 0 failures.

Lmk if there's anything else you want changed.

**Review by zhiyuan1i — APPROVED** [review]
